test/functional/conditional-validation.spec.ts

```diff
@@ -92,4 +92,72 @@ describe('conditional validation', () => {
       expect(errors[0].value).toEqual('bad_value');
     });
   });
+
+  it('should validate a property when condition returns true and value is supplied', function () {
+    class MyClass {
+      @IsOptional(item => item.age === 33)
+      @Equals('test')
+      title: string = 'bad_value';
+      age: number = 33;
+    }
+
+    const model = new MyClass();
+    return validator.validate(model).then(errors => {
+      expect(errors.length).toEqual(1);
+      expect(errors[0].target).toEqual(model);
+      expect(errors[0].property).toEqual('title');
+      expect(errors[0].constraints).toEqual({ equals: 'title must be equal to test' });
+      expect(errors[0].value).toEqual('bad_value');
+    });
+  });
+
+  it('should validate a property when condition returns false and value is supplied', function () {
+    class MyClass {
+      @IsOptional(item => item.age === 34)
+      @Equals('test')
+      title: string = 'bad_value';
+      age: number = 33;
+    }
+
+    const model = new MyClass();
+    return validator.validate(model).then(errors => {
+      expect(errors.length).toEqual(1);
+      expect(errors[0].target).toEqual(model);
+      expect(errors[0].property).toEqual('title');
+      expect(errors[0].constraints).toEqual({ equals: 'title must be equal to test' });
+      expect(errors[0].value).toEqual('bad_value');
+    });
+  });
+
+  it('should validate a property when condition returns true and value is not supplied', function () {
```

Comment:
when the condition returns true (meaning the property is optional), the validation should _not_ happen if the property was not supplied.